sensing/autoware_cuda_utils/include/autoware/cuda_utils/cuda_gtest_utils.hpp

```
@@ -0,0 +1,67 @@
+// Copyright 2022 Tier IV, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This code is licensed under CC0 1.0 Universal (Public Domain).
+// You can use this without any limitation.
+// https://creativecommons.org/publicdomain/zero/1.0/deed.en
+// borrowed from https://proc-cpuinfo.fixstars.com/2019/02/cuda_smart_pointer/
+
+#ifndef AUTOWARE__CUDA_UTILS__CUDA_GTEST_UTILS_HPP_
+#define AUTOWARE__CUDA_UTILS__CUDA_GTEST_UTILS_HPP_
+
+#include <cuda_runtime.h>
+#include <gtest/gtest.h>
+
+#define SKIP_TEST_IF_CUDA_UNAVAILABLE()                              \
+  if (!autoware::cuda_utils::is_cuda_runtime_available()) {          \
+    GTEST_SKIP() << "CUDA runtime is not available. Skipping test."; \
+  }
+
+namespace autoware::cuda_utils
+{
+
+/**
+ * @brief Check if CUDA is available at runtime.
+ * @return true if CUDA is available, false otherwise.
+ */
+inline bool is_cuda_runtime_available()
+{
+  int device_count = 0;
+  cudaError_t error = cudaGetDeviceCount(&device_count);
```

nit:
[cudaGetDeviceCount](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__DEVICE.html#group__CUDART__DEVICE_1g18808e54893cfcaafefeab31a73cc55f) is not thread safe. Colcon (whether CI runner with GPU or local machine) run tests for multiple packages -> First thread calls `cudaGetDeviceCount()` and triggers CUDA runtime initialization -> Second thread calls `cudaGetDeviceCount()` simultaneously and due to race condition might get initialization error `cudaErrorInitializationError`.

I marked it as nit because it will fail silently - test will be skipped.

---

> I read the documentation you shared, but I am not finding info on thread safety.

I probably overinterpreted the lazy loading behavior and wrongly assumed that if there is no mention of thread safety, it is probably not safe. I also couldn't reproduce the error with the generated unit test. Sorry for the trouble! :pepe-shy: